### ISSUE TITLE
WIP: restructuring config

### ISSUE THREAD

**Issue by deargle** [NYUCCL/psiTurk#435]
I'm opening this issue so I can think-aloud and document some of the changes I'm making to the config approach on the master branch.

Right now, I'm doing things like:
* making it so that a config.txt file is not necessary
* totally removing the global config settings. before, they used to store psiturk ad server credentials and aws credentials, but i think it makes more sense to let boto look in the [typical locations](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html) for aws configs
* making it so that every setting can be overridden by env vars.
    * psiturk will have defaults for many (not all) config settings
    * users can create their own config.txt which can set their _preferred_ defaults
    * users can set env vars which [win over any of the above two options](https://stackoverflow.com/a/52972733)

edit: i'm merging https://github.com/NYUCCL/psiTurk/pull/415 in with my wip

also see https://github.com/NYUCCL/psiTurk/issues/391

**Comment by deargle** [comment]
I think I'm going to make it so that env var settings have to be prefixed with `PSITURK_` if they are going to override any config setting -- except for the sole cases of `PORT` and `DATABASE_URL`, which heroku set and expect the app to respect.

**Comment by gureckis** [comment]
on minor comment about this is that using environment variables can be a hurdle to reproducibility.  so like the idea that you can give your experiment code to someone else and they can run the same experiment relatively easily should be one design consideration.  the config.txt makes those setting explicit and if more move into environment variables which might be set by e.g., command line, this might make reproducibility harder.  i *think* your idea of having users have their own config.txt file can still deal with that but should be part of the agenda perhaps.

**Comment by deargle** [comment]
I think we should still be okay. Anything _could_ be set in the config.txt, but overridden via env var when deployed on heroku etc. And, it will be possible for users to specify their env var settings in a `.env` key-value file, which can be loaded by python-dotenv. More latter gotta run rn, but like this:

```python
# prefer environment
        these_as_they_are = ['PORT','DATABASE_URL'] # heroku sets these
        for section, config_vars in configs.items():
            for config_var in config_vars:
                config_var_upper = config_var.upper()
                config_val_env_override = None

                if config_var_upper in these_as_they_are:
                    config_val_env_override = os.environ.get(config_var_upper)

                # prefer any PSITURK_ key even over `these_as_they_are` variants
                psiturk_key = f'PSITURK_{config_var_upper}'
                if psiturk_key in os.environ:
                    config_val_env_override = os.environ.get(psiturk_key)

                if config_val_env_override:
                    config_val_env_override = config_val_env_override.lower()
                    self.set(section, conf_var, config_val_env_override)
```

**Comment by deargle** [comment]
@gureckis following up. Thinking about it, there's certain config settings that don't apply to reproducibility, such as:
* number of server threads
* server debug level and log files
* default shell mode
* ad url
* host and port

there's others that do definitely apply to reproducibility, such as:
* task name and description
* keywords
* number of conditions
* allow_repeats

Then there's others that have no business being in a config file, such as:
* database url
* login_pw and username

For the sake of reproducibility, maybe we flag certain sections that _must_ or _are encouraged_ to be in a config file, or in a README.md of sorts? Then let the others be override-able.

On the other hand, perhaps the only users who will be using env vars are the ones who have different environments, and who need the ability to override certain things. Perhaps most will still use the config.txt for most things, except DATABASE_URL and aws creds etc.

Here's another observation,
* task descriptors such as the task title, keywords, amt settings, and description only need to be available from where the hits are being created, which is _not_ heroku

One more thing, it'd be cool if psiturk logged config settings to a file every time it created a HIT. Because task description and title, payment amount (!!), num_conditions can change between hit postings... I currently keep a manual log of how much I pay for each round (sometimes the task gets longer etc), but currently there's no easy record of how much a given participant was base-paid. Could stash that stuff in the new-ish amt_hit table.

**Comment by deargle** [comment]
I'm using https://github.com/grafana/grafana/blob/master/conf/sample.ini as an example for stylizing and ultra-commenting the config file

**Comment by deargle** [comment]
@gureckis I'm sunsetting all code related to the psiturk ad server in this PR. And by sunset, I mean completely remove :-) -- sound good? It will still be usable via psiturk v2

**Comment by gureckis** [comment]
sure!

todd

**Comment by deargle** [comment]
@gureckis are you on board with sunsetting the experiment exchange server, too? rather than maintaining a database of github links, we could instead move towards having people submit PRs against this repo with links to experiments that they want to share. After all, the experiment exchange code basically just does a `git clone`

I'm playing around with moving psiturk.org into jekyll (basically done after 1hr), which would mean that psiturk.org could be hosted as a github-pages site right out of this repo. No more separate webhosting as ultimate goal. Thoughts?

**Comment by deargle** [comment]
e.g., psiturk.org would live on the `gh-pages` branch, https://github.com/NYUCCL/psiTurk/tree/gh-pages

**Comment by deargle** [comment]
@gureckis I just put all of the experiment exchange listings into the docs, as a single page "experiment exchange." At the top are loose docs for how psiturkers can contribute to the exchange -- by putting their experiment in a github repository, and submitting a PR to add it to our docs. I also suggest that if they need help getting it into the docs, they can post it as an issue (or to the google group, I should add) and we'll help them get it in.

I also did a _lot_ of housekeeping on the docs. I'm novice with sphinx now. You can see a build of the docs that I want to merge into master [here](https://psiturk.readthedocs.io/en/config-refactor/) -- I think with these docs, we can release v3.0.0 beta 🎉 -- can you peek at the docs?

**Comment by Portll** [comment]
I'm not the most active on the group but I like the look of what's been
done and it definitely appears more maintainable, easier to understand and
somewhat more coherent. Docs are definitely a huge improvement.

Good work Dave (and Todd)

>

**Comment by deargle** [comment]
Thanks @Portll! I just realized that I had a proofreading commit that hadn't been pushed -- it's pushed and new docs are building now. The most substantial changes in this push is that I made some of the FAQs more useful, that I reorganized the "databases_overview" a bit, and I did some formatting and link-fixing here and there. But if you think the pre-proofread version was already an improvement, then 😄 this will only make it better!
